packages/contrast-colors/index.js

```diff
@@ -13,7 +13,500 @@ governing permissions and limitations under the License.
+      if(!color.ratios) throw new Error(`Color ${color.name}'s ratios are undefined`);
+    });
+    if (!colorSpaces[this._output]) {
+      throw new Error(`Output “${colorspace}” not supported`);
+    }
+    
+    this._setContrasts(this._contrast);
+    this._findContrastColors();
+    this._findContrastColorValues();
+  }
+  
+  set contrast(contrast) {
+    this._contrast = contrast;
+    this._setContrasts(contrast);
+    this._findContrastColors();
+  }
+  get contrast() {
+    return this._contrast;
+  }
+
+  set lightness(lightness) {
+    this._lightness = lightness;
+    this._setBackgroundColor(this._backgroundColor);
+    this._findContrastColors();
+  }
+  get lightness() {
+    return this._lightness;
+  }
+  
+  set backgroundColor(backgroundColor) {
+    this._setBackgroundColor(backgroundColor);
+    this._findContrastColors();
+  }
+  get backgroundColorValue() {
+    return this._backgroundColorValue;
+  }
+  get backgroundColor() {
+    return this._backgroundColor;
+  }
+  
+  // Add a getter and setter for colors
+  set colors(colors) {
+    this._colors = colors;
+    this._findContrastColors();
+  }
+  get colors() {
+    return this._colors;
+  }
+
+  set output(output) {
+    this._output = output;
+    this._colors.forEach(element => {
+      element.output = this._output;
+    });
+    this._backgroundColor.output = this._output;
+
+    this._findContrastColors();
+  }
+  get output() {
+    return this._output;
+  }
+  
+  get contrastColors() {
+    return this._contrastColors;
+  }
+
+  get contrastColorValues() {
+    return this._contrastColorValues;
+  }
+
+  _setBackgroundColor(backgroundColor) {
+    if(typeof backgroundColor === 'string') {
+      // If it's a string, convert to Color object and assign lightness.
+      const newBackgroundColor = new BackgroundColor({name: 'background', colorKeys: [backgroundColor], output: 'RGB'});
+      const calcLightness = Number((d3.hsluv(backgroundColor).v).toFixed());
+
+      return this._backgroundColor = newBackgroundColor, this._lightness = calcLightness, this._backgroundColorValue = newBackgroundColor[this._lightness];
+      // console.log(`String background color of ${backgroundColor} converted to ${newBackgroundColor}`)
+    } else {
+      // console.log(`NOT a string for background, instead it is ${JSON.stringify(backgroundColor)}`)
+      backgroundColor.output = 'RGB';
+      const calcBackgroundColorValue = backgroundColor.backgroundColorScale[this._lightness];
+
+      // console.log(`Object background \nLightness: ${this._lightness} \nBackground scale: ${backgroundColor.backgroundColorScale}\nCalculated background value of ${calcBackgroundColorValue}`)
+      return this._backgroundColor = backgroundColor, this._backgroundColorValue = calcBackgroundColorValue;
+    }
+  }
+
+  _setBackgroundColorValue() {
+    return this._backgroundColorValue = this._backgroundColor.backgroundColorScale[this._lightness];
+  }
+
+  _setContrasts(contrastMultiplier) {
```

Comment:
This entire function approach needs to be revisited in order for adaptability of themes to be realized. This current function re-maps the existing contrast values by passing through a multiplier function. This effectively eliminates any context to the original values, since they are directly set to the original color objects with a setter function.

This is problematic in real-world scenarios where contrast is dynamically changed. A change in contrast is multiplied against *the previous changed value* rather than multiplying against the original contrast values. For example:
```
// original    multiplier.  output
3.5.         x 1.2           =   4.  /// ((3.5 -1) x 1.2) +1
```
Now if we don't retain the original value and contrast is decreased to 1.1, we see this:
```
4             x 1.1            = 4.3
```
In this scenario, even though contrast was relatively _decreased_, since the mapping directly manipulates the ratios of the Color objects, they cannot be manipulated in a way that makes sense.

This function should somehow create a new temporary set of ratios or color objects that the output `contrastColors` are generated based on.

Reply:
@lazd I'd love to get your thoughts on this one.

Reply:
Nevermind I figured this out -- I removed this function altogether and just apply the multiplier function to each contrast ratio during the `_findContrastColors` function, essentially only applying it when I need it and when it's contextual -- never storing the "new" ratios to the state. Problem solved.